R/forecastingdata_org.R

```diff
@@ -0,0 +1,380 @@
+  LOW_FREQ_VALS <- c("4 sec", "1 min", "10 min", "30 min", "1 hour")
+  HIGH_FREQUENCIES <- c("daily", "weekly", "monthly", "quarterly", "yearly")
+  HIGH_FREQ_VALS <- c("1 day", "1 week", "1 month", "3 months", "1 year")
+  FREQUENCIES <- c(LOW_FREQUENCIES, HIGH_FREQUENCIES)
+  FREQ_VALS <- c(LOW_FREQ_VALS, HIGH_FREQ_VALS)
+  
+  
+  # Create a hashmap containing possible frequency key-value pairs
+  FREQ_MAP <- list()
+  
+  for(f in seq_along(FREQUENCIES))
+    FREQ_MAP[[FREQUENCIES[f]]] <- FREQ_VALS[f]
+  
+  
+  if(is.character(file)) {
+    file <- file(file, "r")
+    on.exit(close(file))
+  }
+  if(!inherits(file, "connection"))
+    stop("Argument 'file' must be a character string or connection.")
+  if(!isOpen(file)) {
+    open(file, "r")
+    on.exit(close(file))
+  }
+  
+  # Read meta-data
+  col_names <- NULL
+  col_types <- NULL
+  frequency <- NULL
+  forecast_horizon <- NULL
+  contain_missing_values <- NULL
+  contain_equal_length <- NULL
+  index_var <- NULL
+  
+  line <- readLines(file, n = 1) # n is no: of lines to read
+  
+  while(length(line) && regexpr('^[[:space:]]*@data', line, perl = TRUE) == -1) { # Until read @data, run this loop (-1 indicate no match with the regular expression yet)
+    
+    if(regexpr('^[[:space:]]*@', line, perl = TRUE) > 0) { # This condition will be true for lines starting with @
+      
+      con <- textConnection(line)
+      line <- scan(con, character(), quiet = TRUE) # Creating a vector containing the words in a line (ex: "@attribute" "series_name" "string")
+      close(con)
+      
+      if(line[1] == "@attribute"){
+        if(length(line) != 3)  # Attributes have both name and type
+          stop("Invalid meta-data specification.")
+        
+        if(is.null(index) & line[3] == "date") # Searching for a valid index, if index is not given 
+          index_var <- line[2]
+        
+        col_names <- c(col_names, line[2])
+        col_types <- c(col_types, line[3])
+      }else{
+        if(length(line) != 2) # Other meta-data have only values
+          stop("Invalid meta-data specification.")
+        
+        if(line[1] == "@frequency")
+          frequency <- line[2]
+        else if(line[1] == "@horizon")
+          forecast_horizon <- as.numeric(line[2])
+        else if(line[1] == "@missing")
+          contain_missing_values <- as.logical(line[2])
+        else if(line[1] == "@equallength")
+          contain_equal_length <- as.logical(line[2])
+      }
+    }
+    line <- readLines(file, n = 1)
+  }
+  
+  if(length(line) == 0)
+    stop("Missing data section.")
+  if(is.null(col_names))
+    stop("Missing attribute section.")
+  
+  line <- readLines(file, n = 1)
+  
+  if(length(line) == 0)
+    stop("Missing series information under data section.")
+  
+  for(col in col_names)
+    assign(col, NULL)
+  
+  values <- NULL
+  row_count <- 0
+  
+  # Get data
+  while(length(line) != 0){
```

Comment:
Could this be done with functional programming? How is each line processed, perhaps there are other functions which can be used to do most of the processing?